mono/metadata/oop.c

```diff
@@ -0,0 +1,422 @@
+    return left;
+}
+
+static int
+oop_jit_info_table_chunk_index(
+    const MonoJitInfo** chunk_data,
+    int num_elements,
+    const gint8 *addr)
+{
+    const MonoJitInfo* ji;
+    int left = 0, right = num_elements;
+
+    while (left < right) {
+        int pos = (left + right) / 2;
+
+        const gint8 *code_start;
+        const gint8 *code_end;
+        int code_size;
+
+        ji = chunk_data[pos];
+
+        code_start = (const gint8*)read_pointer(OFFSET_MEMBER(MonoJitInfo, ji, code_start));
+        code_size = read_dword(OFFSET_MEMBER(MonoJitInfo, ji, code_size));
+        code_end = code_start + code_size;
+
+        if (addr < code_end)
+            right = pos;
+        else
+            left = pos + 1;
+    }
+
+    g_assert(left == right);
+
+    return left;
+}
+
+static const MonoJitInfo*
+oop_jit_info_table_find(
+    const MonoDomain *domain,
+    const char *addr)
+{
+    const MonoJitInfoTable* tablePtr;
+    const MonoJitInfoTableChunk** chunkListPtr;
+    MonoJitInfoTableChunk chunk;
+    MonoJitInfo ji;
+    int chunk_pos, pos;
+
+    // Get the domain's jit_info_table pointer.
+    tablePtr = read_pointer(OFFSET_MEMBER(MonoDomain, domain, jit_info_table));
+    if (tablePtr == NULL)
+        return NULL;
+
+    int num_chunks = read_dword(OFFSET_MEMBER(MonoJitInfoTable, tablePtr, num_chunks));
+
+    // Get the chunk array
+    chunkListPtr = (const MonoJitInfoTableChunk**)OFFSET_MEMBER(MonoJitInfoTable, tablePtr, chunks);
+
+    chunk_pos = oop_jit_info_table_index(chunkListPtr, num_chunks, (const gint8*)addr);
+    if (chunk_pos > num_chunks)
+        return NULL;
+    
+    // read the entire chunk
+    read_memory(&chunk, read_pointer(chunkListPtr + chunk_pos), sizeof(MonoJitInfoTableChunk));
+
+    pos = oop_jit_info_table_chunk_index((const MonoJitInfo**)chunk.data, chunk.num_elements, addr);
+    if (pos > chunk.num_elements)
+        return NULL;
+
+    /* We now have a position that's very close to that of the
+    first element whose end address is higher than the one
+    we're looking for.  If we don't have the exact position,
+    then we have a position below that one, so we'll just
+    search upward until we find our element. */
+    do {
+        read_memory(&chunk, read_pointer(chunkListPtr + chunk_pos), sizeof(MonoJitInfoTableChunk));
+
+        while (pos < chunk.num_elements) {
+            read_memory(&ji, chunk.data[pos], sizeof(ji));
+
+            ++pos;
+
+            if (ji.d.method == NULL) {
+                continue;
+            }
+            if ((gint8*)addr >= (gint8*)ji.code_start
+                && (gint8*)addr < (gint8*)ji.code_start + ji.code_size) {
+                return chunk.data[pos];
+            }
+
+            /* If we find a non-tombstone element which is already
+            beyond what we're looking for, we have to end the
+            search. */
+            if ((gint8*)addr < (gint8*)ji.code_start)
+                goto not_found;
```

Comment:
Agreed. This is a clone of jit_info_table_find (adapted to be OOP) and I missed that. As a result of this, I've also added comments to both functions that one is derived from the other.